### PR TITLE
os/bluestore: dynamic CF configuration; put pglog omap in separate CF

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by markhpc** [comment]
I only scanned this briefly so please feel free to disregard if I'm smoking crack.  This feels a bit off to me.  I don't like leaking the RocksDB column family implementation details  out into bluestore via the KeyValueDB interface unless we have a really good reason (Maybe we do!).  I guess I'd favor keeping this out of bluestore and simply having configurables at the rocksdb KeyvalueDB level to put different prefixes in different column families (or even shard the DB or whatever).  Personally I want the KeyValueDB to look like as much of a black box as possible (that ultimately we still control, but bluestore knows nothing about other than the simplest API necessary for storing keys/values).

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -10914,8 +10955,10 @@ int BlueStore::_omap_setheader(TransContext *txc,
   } else {
     txc->note_modified_object(o);
   }
+  const string& prefix =
+    o->onode.is_pgmeta_omap() ? PREFIX_PGMETA_OMAP : PREFIX_OMAP;
```
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.cc]: Need to set **FLAG_PGMETA_OMAP** first, see `BlueStore::_omap_setkeys`

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -11052,14 +11104,19 @@ int BlueStore::_clone(TransContext *txc,
...
-    _do_omap_clear(txc, newo->onode.nid);
+    _do_omap_clear(txc,
+		   newo->onode.is_pgmeta_omap() ? PREFIX_PGMETA_OMAP
+		   : PREFIX_OMAP,
+		   newo->onode.nid);
   }
   if (oldo->onode.has_omap()) {
     dout(20) << __func__ << " copying omap data" << dendl;
     if (!newo->onode.has_omap()) {
       newo->onode.set_omap_flag();
```
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.cc]: Need to set **FLAG_PGMETA_OMAP** too.  Do we really get the chance to ever clone a pgmeta object?

File: src/os/bluestore/bluestore_types.h [hunk 3]
```diff
@@ -952,7 +952,8 @@ struct bluestore_onode_t {
   uint8_t flags = 0;
 
   enum {
-    FLAG_OMAP = 1,
+    FLAG_OMAP = 1,       ///< object may have omap data
+    FLAG_PGMETA_OMAP = 2,  ///< omap data is in meta omap prefix
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/os/bluestore/bluestore_types.h]: we have to distinguish between existing objects (in the normal omap space) and new objects.

[the remark this answers is not in the document]
Comment (xiexingguo) [reply src/os/bluestore/bluestore_types.h]: Ahh, right.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-10-13_14:38:14-rados-wip-sage2-testing-2017-10-13-0719-distro-basic-smithi/

**Pull request merged**